R/vector.R

```diff
@@ -0,0 +1,213 @@
+#' str(list(inputs, arg3 = c("c1", "c2")))
+#' str(c(inputs, arg3 = c("c1", "c2")))
+splice <- function(...) {
+  dots <- list(...)
+  if (length(dots) == 0) {
+    return(dots)
+  }
+
+  names <- Map(function(dot, name) {
+    if (is_bare_list(dot)) {
+      names2(dot)
+    } else {
+      name
+    }
+  }, dots, names2(dots))
+  names <- unlist(names, recursive = FALSE)
+
+  is_not_list <- vapply_lgl(dots, function(x) !is_bare_list(x))
+  dots[is_not_list] <- lapply(dots[is_not_list], list)
+
+  out <- unlist(dots, recursive = FALSE)
+  set_names(out, names)
+}
+
+
+#' Prepend a vector.
+#'
+#' This is a companion to \code{\link[base]{append}()} to help merging
+#' two lists or atomic vectors. \code{prepend()} is a clearer semantic
+#' signal than `c()` that a vector is to be merged at the beginning of
+#' another, especially in a pipe chain.
+#'
+#' @param x the vector to be modified.
+#' @param values to be included in the modified vector.
+#' @param before a subscript, before which the values are to be appended.
+#' @return A merged vector.
+#' @export
+#' @examples
+#' x <- as.list(1:3)
+#'
+#' append(x, "a")
+#' prepend(x, "a")
+#' prepend(x, list("a", "b"), before = 3)
+prepend <- function(x, values, before = 1) {
+  n <- length(x)
+  stopifnot(before > 0 && before <= n)
+
+  if (before == 1) {
+    c(values, x)
+  } else {
+    c(x[1:(before - 1)], values, x[before:n])
+  }
+}
+
+#' Modify a vector.
+#'
+#' This function merges a list of arguments into a vector. It always
+#' returns a list.
+#'
+#' @param .x A vector to modify.
+#' @param ...,.elts List of elements to merge into \code{.x}.
+#' @return A modified vector upcasted to a list.
+#' @export
+#' @examples
+#' modify(c(1, b = 2, 3), 4, b = "foo")
+#'
+#' x <- list(a = 1, b = 2)
+#' y <- list(b = 3, c = 4)
+#' modify(x, .elts = y)
+modify <- function(.x, ..., .elts = list()) {
+  out <- as.list(.x)
+  args <- c(list(...), .elts)
+
+  args_nms <- names(args)
+  exists <- has_names(args) & args_nms %in% names(out)
+
+  for (nm in args_nms[exists]) {
+    out[[nm]] <- args[[nm]]
+  }
+
+  c(out, args[!exists])
+}
+
+
+#' Helper to create vectors with matching length.
+#'
+#' These functions take the idea of \code{\link{seq_along}} and
+#' generalise it to creating lists (\code{list_along}) and repeating
+#' values (\code{rep_along}). The dots are forwarded to
+#' \code{\link[base]{structure}()} to make it easy to add attributes.
+#'
+#' @inheritParams with_attributes
+#' @param .x A vector.
+#' @param .y Values to repeat.
+#' @examples
+#' x <- 0:5
+#' rep_along(x, 1:2)
+#' rep_along(x, 1)
+#' list_along(x)
+#'
+#' # You can also add attributes with additional arguments:
+#' rep_along(x, 1, class = "my_class")
+#' double_along(x, class = "my_class")
+#' @name along
+#' @seealso new-vectors
+NULL
+
+#' @export
+#' @rdname along
+logical_along <- function(.x, ..., .attrs = list()) {
+  with_attributes(vector("logical", length(.x)), ..., .attrs = .attrs)
```

Comment:
I don't love the default values that `vector` fills in. I'd rather have a generic `na_along(x)` which would take the type from `x`, optionally overridden by a "type" argument (or _maybe_ `na_dbl_along(x)` etc).